pkg/datasource/plugin.go

```diff
@@ -0,0 +1,219 @@
+	Group            []string `json:"group"`
+	MetricColumn     string   `json:"metricColumn"`
+	OrderByCol       string   `json:"orderByCol"`
+	OrderBySort      string   `json:"orderBySort"`
+	Partitioned      bool     `json:"partitioned"`
+	PartitionedField string   `json:"partitionedField"`
+	Project          string   `json:"project"`
+	RawQuery         bool     `json:"rawQuery"`
+	RawSQL           string   `json:"rawSql"`
+	RefID            string   `json:"refId"`
+	Sharded          bool     `json:"sharded"`
+	Table            string   `json:"table"`
+	TimeColumn       string   `json:"timeColumn"`
+	TimeColumnType   string   `json:"timeColumnType"`
+	Location         string   `json:"location"`
+}
+
+type instanceSettings struct {
+	httpClient *http.Client
+}
+
+func newDataSourceInstance(_ backend.DataSourceInstanceSettings) (instancemgmt.Instance, error) {
+	return &instanceSettings{
+		httpClient: &http.Client{},
+	}, nil
+}
+
+// Dispose is called before creating a new instance to allow plugin authors to cleanup
+func (s *instanceSettings) Dispose() {}
+
+// BigQueryDatasource is an example datasource used to scaffold
+// new datasource plugins with an backend.
+type BigQueryDatasource struct {
+	im instancemgmt.InstanceManager
+}
+
+// New creates a datasource that is sued for querying BigQuery
+func New() datasource.ServeOpts {
+	im := datasource.NewInstanceManager(newDataSourceInstance)
+	ds := &BigQueryDatasource{
+		im: im,
+	}
+
+	return datasource.ServeOpts{
+		QueryDataHandler:   ds,
+		CheckHealthHandler: ds,
+	}
+}
+
+const maxParallelQueries = 20
+
+// QueryData handles multiple queries and returns multiple responses.
+// req contains the queries []DataQuery (where each query contains RefID as a unique identifier).
+// The QueryDataResponse contains a map of RefID to the response for each query, and each response
+// contains Frames ([]*Frame).
+func (bq *BigQueryDatasource) QueryData(ctx context.Context, req *backend.QueryDataRequest) (*backend.QueryDataResponse, error) {
+	response := backend.NewQueryDataResponse()
+	mtx := &sync.Mutex{}
+	wg := &sync.WaitGroup{}
+	queries := req.Queries
+
+	log.DefaultLogger.Info("Running %d queries", len(queries))
+	// Execute the queries in parallel, up to maxParallelQueries, and collect the responses.
+	for {
+		var next = maxParallelQueries
+		if llen := len(queries); llen < maxParallelQueries {
+			next = llen
+		}
+		log.DefaultLogger.Warn("Running %d queries", next)
+
+		wg.Add(next)
+		qctx, cancel := context.WithTimeout(ctx, time.Minute)
+		for _, q := range queries[:next] {
+			go func(dq backend.DataQuery) {
+				//res := bq.doQuery(qctx, dq)
+				if !errors.Is(qctx.Err(), context.DeadlineExceeded) {
+					mtx.Lock()
+					response.Responses[dq.RefID] = backend.DataResponse{}
+					mtx.Unlock()
+				}
+				wg.Done()
+			}(q)
+		}
+		wg.Wait()
+		cancel()
+		if next != maxParallelQueries {
+			break
+		}
+		queries = queries[next:]
+	}
+
+	return response, nil
+}
```

Comment:
So the way to do this is not batches of parallel queries, but instead use a counting semaphore to only allow N requests in flight at once - decrement the count when a request completes and increment when starting a request, blocking when the limit is reached.

Reply:
https://fodor.org/blog/go-counting-semaphores/ is an example using a buffered channel as a counting semaphore.